From 083f0109fff94d2515a537d71d6b534efbd07799 Mon Sep 17 00:00:00 2001
From: Jeff Xu <jeffxu@google.com>
Date: Fri, 19 Aug 2022 20:52:34 +0000
Subject: [PATCH 1/3] CHROMIUM: chromeOS LSM to block memfd execution.
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

In chromeOS, memfd is frequently seen as a tool
for noexec bypass. This noexec bypass enables
“Confused deputy attack” for root escalation.

Example:
https://crbug.com/1305267
https://crbug.com/1336668

This change blocks memfd execution in chromeOS LSM.

Existing hook to use: security_bprm_creds_for_exec

In order to block user space memfd execution, we need
a hook in bprm_execve() after do_open_execat() call.

bprm_execve() is a common code path from 3 entry points:
1> syscall: execve
2> syscall: execveat
3> kernel_execve (used by kernel to invoke user program)

do_open_execat() gets struct file pointer from fd or filename,
after permission checks.

Since the memfd check is based on file pointer,
and we want to exit as early as possible, before loading
the executable,  it is ideal to have the check right after
do_open_execat call, however that be a new hook.

After do_open_execat(), there is call to sched_exec(), then
hook security_bprm_creds_for_exec(), sched_exec() is a
balancing opportunity, (so not too much cost), so we could use
security_bprm_creds_for_exec() for our purpose.

The file pointer is coming from bprm->file.

With this, chromeOS LSM will block all user space memfd execution.

BUG=b:242725803
TEST=run manual test

Signed-off-by: Jeff Xu<jeffxu@google.com>
Change-Id: Ib261d76973e91a29916c05bf0ecf51d06a28be07
Reviewed-on: https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/3834992
Reviewed-by: Guenter Roeck <groeck@chromium.org>
Reviewed-by: Jorge Lucangeli Obes <jorgelo@chromium.org>

[rebase61(tzungbi):
    Squashed:
	FIXUP: CHROMIUM: chromeOS LSM to block memfd execution.
	CHROMIUM: adds audit logging for memfd execution
	Part of "CHROMIUM: Add logging for lsm"
]
Signed-off-by: Tzung-Bi Shih <tzungbi@chromium.org>

Change-Id: Id11b52d48cd0feba8dcb41ef39e15c0ddefbc7e6
---

Notes:
    42f219fd81cc02aa8332fb33d8661d099165a1b1
    8316d9884d7bbb9e5f29be1210b25c7e1fe8c5bd
    549e1b15b2061dddf55175f2c35d721502b5593b
    6c54efd362474aeabc995d9fc351167565027da9
    74705887564932e4269972c9ebba754faa066363
    9f12306c254bfeae7fb1d5e730f47deb2652ec9d

 security/chromiumos/lsm.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/security/chromiumos/lsm.c b/security/chromiumos/lsm.c
index 3bf4a9d75f4b..b64009523a31 100644
--- a/security/chromiumos/lsm.c
+++ b/security/chromiumos/lsm.c
@@ -20,6 +20,8 @@
 #define pr_fmt(fmt) "Chromium OS LSM: " fmt
 
 #include <asm/syscall.h>
+#include <linux/audit.h>
+#include <linux/binfmts.h>
 #include <linux/cred.h>
 #include <linux/fs.h>
 #include <linux/fs_parser.h>
@@ -33,6 +35,7 @@
 #include <linux/sched/task_stack.h>
 #include <linux/sched.h>	/* current and other task related stuff */
 #include <linux/security.h>
+#include <linux/shmem_fs.h>
 #include <uapi/linux/mount.h>
 
 #include "inode_mark.h"
@@ -252,11 +255,34 @@ static int chromiumos_sb_eat_lsm_opts(char *options, void **mnt_opts)
 	return 0;
 }
 
+static int chromiumos_bprm_creds_for_exec(struct linux_binprm *bprm)
+{
+	struct file *file = bprm->file;
+
+	if (shmem_file(file)) {
+		char *cmdline = printable_cmdline(current);
+
+		audit_log(
+			audit_context(),
+			GFP_ATOMIC,
+			AUDIT_AVC,
+			"ChromeOS LSM: memfd execution attempt, cmd=%s, pid=%d",
+			cmdline ? cmdline : "(null)",
+			task_pid_nr(current));
+		kfree(cmdline);
+
+		pr_notice_ratelimited("memfd execution blocked\n");
+		return -EACCES;
+	}
+	return 0;
+}
+
 static struct security_hook_list chromiumos_security_hooks[] = {
 	LSM_HOOK_INIT(sb_mount, chromiumos_security_sb_mount),
 	LSM_HOOK_INIT(inode_follow_link, chromiumos_security_inode_follow_link),
 	LSM_HOOK_INIT(file_open, chromiumos_security_file_open),
 	LSM_HOOK_INIT(sb_eat_lsm_opts, chromiumos_sb_eat_lsm_opts),
+	LSM_HOOK_INIT(bprm_creds_for_exec, chromiumos_bprm_creds_for_exec),
 };
 
 static int __init chromiumos_security_init(void)
-- 
2.39.0.314.g84b9a713c41-goog


From 434dd002db26b0e9f12f093b72c7488a2cfb953e Mon Sep 17 00:00:00 2001
From: Jeff Xu <jeffxu@google.com>
Date: Fri, 7 Oct 2022 19:49:15 +0000
Subject: [PATCH 2/3] CHROMIUM: chromeOS LSM to block bpf_write_user

BPF can overwrite user memory and was used in root escalation.
Example: https://crbug.com/1320051

Currently ChromeOS doesn't use bpf_write_user, this LSM blocks
BPF programs that uses bpf_write_user.

BUG=b:251868129
TEST=load a BPF program with bpf_write_user, it is blocked.

Signed-off-by: Jeff Xu <jeffxu@google.com>
Change-Id: Ie48c93b775e45c5fcb05133f874758590fadd86f
Reviewed-on: https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/3939652
Reviewed-by: Guenter Roeck <groeck@chromium.org>

[rebase61(tzungbi):
    Squashed:
	Part of "CHROMIUM: Add logging for lsm"
]
Signed-off-by: Tzung-Bi Shih <tzungbi@chromium.org>

Change-Id: Iae3cadf99fc3828a5f7789eae73b7d28f6258943
---
 security/chromiumos/lsm.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/security/chromiumos/lsm.c b/security/chromiumos/lsm.c
index b64009523a31..8956ddf9c1b2 100644
--- a/security/chromiumos/lsm.c
+++ b/security/chromiumos/lsm.c
@@ -277,12 +277,23 @@ static int chromiumos_bprm_creds_for_exec(struct linux_binprm *bprm)
 	return 0;
 }
 
+static int chromiumos_locked_down(enum lockdown_reason what)
+{
+	if (what == LOCKDOWN_BPF_WRITE_USER) {
+		pr_notice_ratelimited("BPF_WRITE_USER blocked\n");
+		return -EACCES;
+	}
+
+	return 0;
+}
+
 static struct security_hook_list chromiumos_security_hooks[] = {
 	LSM_HOOK_INIT(sb_mount, chromiumos_security_sb_mount),
 	LSM_HOOK_INIT(inode_follow_link, chromiumos_security_inode_follow_link),
 	LSM_HOOK_INIT(file_open, chromiumos_security_file_open),
 	LSM_HOOK_INIT(sb_eat_lsm_opts, chromiumos_sb_eat_lsm_opts),
 	LSM_HOOK_INIT(bprm_creds_for_exec, chromiumos_bprm_creds_for_exec),
+	LSM_HOOK_INIT(locked_down, chromiumos_locked_down),
 };
 
 static int __init chromiumos_security_init(void)
-- 
2.39.0.314.g84b9a713c41-goog


From 7b7628d5147547d6e2262771277cd373a9f461db Mon Sep 17 00:00:00 2001
From: Jeff Xu <jeffxu@google.com>
Date: Sat, 8 Oct 2022 01:40:01 +0000
Subject: [PATCH 3/3] CHROMIUM: restrict bpf to secagentd only

Restrict BPF to secagentd only.
This is a short term solution till ChromeOS secures
BPF to more processes.

BUG=b:251196147
TEST=manual test, non secagentd is blocked

Signed-off-by: Jeff Xu <jeffxu@google.com>
Change-Id: I2646e2370ee88e08a29e4378bc0767ff21310f26
Reviewed-on: https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/3938901
Reviewed-by: Guenter Roeck <groeck@chromium.org>
Reviewed-by: Aashay Shringarpure <aashay@google.com>

[rebase61(tzungbi):
    Squashed:
	CHROMIUM: ChromeOS LSM: add check for BPF_SYSCALL
	Part of "CHROMIUM: Add logging for lsm"
]
Signed-off-by: Tzung-Bi Shih <tzungbi@chromium.org>

Change-Id: I187b602c541e77678efca558bf715cb0029b6eff
---
 security/chromiumos/lsm.c | 30 ++++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/security/chromiumos/lsm.c b/security/chromiumos/lsm.c
index 8956ddf9c1b2..c8c89fa32056 100644
--- a/security/chromiumos/lsm.c
+++ b/security/chromiumos/lsm.c
@@ -27,6 +27,7 @@
 #include <linux/fs_parser.h>
 #include <linux/fs_struct.h>
 #include <linux/lsm_hooks.h>
+#include <linux/mm.h>
 #include <linux/module.h>
 #include <linux/mount.h>
 #include <linux/namei.h>	/* for nameidata_get_total_link_count */
@@ -287,6 +288,32 @@ static int chromiumos_locked_down(enum lockdown_reason what)
 	return 0;
 }
 
+#ifdef CONFIG_BPF_SYSCALL
+
+static const char secagentd[] = "/usr/sbin/secagentd";
+
+static int chromiumos_bpf(int cmd, union bpf_attr *attr, unsigned int size)
+{
+	char buf[128];
+	int res;
+	int len;
+
+	len = strlen(secagentd);
+	res = get_cmdline(current, buf, sizeof(buf));
+	if (res > 0 && buf[res - 1] == '\0') {
+		// null terminated.
+		res = res - 1;
+	}
+
+	if (res < len || strncmp(buf, secagentd, len)) {
+		pr_notice_ratelimited("bpf syscall blocked\n");
+		return -EACCES;
+	}
+
+	return 0;
+}
+#endif
+
 static struct security_hook_list chromiumos_security_hooks[] = {
 	LSM_HOOK_INIT(sb_mount, chromiumos_security_sb_mount),
 	LSM_HOOK_INIT(inode_follow_link, chromiumos_security_inode_follow_link),
@@ -294,6 +321,9 @@ static struct security_hook_list chromiumos_security_hooks[] = {
 	LSM_HOOK_INIT(sb_eat_lsm_opts, chromiumos_sb_eat_lsm_opts),
 	LSM_HOOK_INIT(bprm_creds_for_exec, chromiumos_bprm_creds_for_exec),
 	LSM_HOOK_INIT(locked_down, chromiumos_locked_down),
+#ifdef CONFIG_BPF_SYSCALL
+	LSM_HOOK_INIT(bpf, chromiumos_bpf),
+#endif
 };
 
 static int __init chromiumos_security_init(void)
-- 
2.39.0.314.g84b9a713c41-goog

